### PR TITLE
[SPARK-23381][CORE] Murmur3 hash generates a different value from other implementations

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
Murmur3 hash generates a different value from the original and other implementations (like Scala standard library and Guava or so) when the length of a bytes array is not multiple of 4.

## How was this patch tested?
Added a unit test.

**Note: When we merge this PR, please give all the credits to Shintaro Murakami.** 

Author: Shintaro Murakami <mrkm4ntr@gmail.com>

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @hvanhovell @sameeragarwal @jkbradley @cloud-fan

File: common/sketch/src/main/java/org/apache/spark/util/sketch/Murmur3_x86_32.java [hunk 1]
```diff
@@ -60,6 +60,8 @@ public static int hashUnsafeWords(Object base, long offset, int lengthInBytes, i
   }
 
   public static int hashUnsafeBytes(Object base, long offset, int lengthInBytes, int seed) {
+    // This is not compatible with original and another implementations.
+    // But remain it for backward compatibility for the components existing before 2.3.
```
Comment (hvanhovell) [inline common/sketch/src/main/java/org/apache/spark/util/sketch/Murmur3_x86_32.java]: NIT: touch this up. The sentence is a bit weird, i.e.: `However we retain this implementation for backwards compatibility with pre-existing (pre 2.3) components.`

File: common/unsafe/src/test/java/org/apache/spark/unsafe/hash/Murmur3_x86_32Suite.java [hunk 2]
```diff
@@ -51,6 +53,23 @@ public void testKnownLongInputs() {
     Assert.assertEquals(-2106506049, hasher.hashLong(Long.MAX_VALUE));
   }
 
+  // SPARK-23381 Check whether the hash of the byte array is the same as another implementations
```
Comment (hvanhovell) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/hash/Murmur3_x86_32Suite.java]: We could add a randomized (disabled) test here. That might increase the confidence we have in the new hash.

**Review by hvanhovell — APPROVED** [review]
LGTM - pending jenkins

**Comment by jkbradley** [comment]
The ML changes LGTM.  Thanks!

**Comment by viirya** [comment]
LGTM

File: common/unsafe/src/test/java/org/apache/spark/unsafe/hash/Murmur3_x86_32Suite.java (hunk 2 above)
Comment (gatorsmile) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/hash/Murmur3_x86_32Suite.java]: Yeah, we can do it in the follow-up PR.

File: common/sketch/src/main/java/org/apache/spark/util/sketch/Murmur3_x86_32.java (hunk 1 above)
Comment (gatorsmile) [reply common/sketch/src/main/java/org/apache/spark/util/sketch/Murmur3_x86_32.java]: Will correct it in the follow-up PR.

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

**Pull request closed without merging**